### ISSUE TITLE
omprog: start program on rsyslog startup

### ISSUE THREAD

**Issue by lego12239** [rsyslog/rsyslog#2776]
It would be conveniently if omprog module has some option(like delayStart, for example) to start specified program on rsyslog startup instead of first arrived message. This is needed in some configurations.

**Comment by rgerhards** [comment]
It would be good to have an explanation of why this may be needed. I admit I have no idea for a use case.

**Comment by lego12239** [comment]
On Thu, Jun 14, 2018 at 01:53:11AM -0700, Rainer Gerhards wrote:
> It would be good to have an explanation of why this may be needed. I admit I have no idea for a use case.

My use case is a fail2ban replacement that run with help of omprog module
and thus be faster than reading log files from disk(I'm sure that it would
even faster if i write a needed rsyslog module, but i have no experience
in this :-); may be later).

Since this is a ban program, it use iptables and must start working
after iptables system-wide rules are restored on system startup. If it start
before iptables start script, then all rules that my program add to iptables
are replaced by default startup settings.

Various distro may have different order of rsyslog and iptables init scripts.
And, if omprog start a program on rsyslog startup, my init script, which
start after rsyslog and iptables init script, can simply send SIGUSR to my
ban program. After what it add needed iptables rules and start working in
a usual manner(before this it simply discard all input).

With action by a signal i can selectively start needed instance of my ban program, if
have many ones; with activation by a log message this would be harder to achieve.

-- 
Олег Неманов (Oleg Nemanov)

**Comment by rgerhards** [comment]
I understand that you just want to start a program when rsyslog starts up. It's not really meant to receive messages from rsyslog? Is that right?

Isn't that a functionality that you could easier implement via systemd? I mean if you craft a unit file that depends on rsyslog or even network start (assuming that iptables is up at that stage.

**Comment by lego12239** [comment]
On Thu, Jun 14, 2018 at 04:14:30AM -0700, Rainer Gerhards wrote:
> I understand that you just want to start a program when rsyslog starts up. It's not really meant to receive messages from rsyslog? Is that right?

  In my case yes.

> Isn't that a functionality that you could easier implement via systemd? I mean if you craft a unit file that depends on rsyslog or even network start (assuming that iptables is up at that stage.

I'm not use systemd :-). But yes, i can make an init script with dependency
of rsyslog and iptables. And in this script i would place a "kill -USR1 `cat /var/run/my.pid`"
which switch the program already run by rsyslog to working mode.

-- 
Олег Неманов (Oleg Nemanov)

**Comment by rgerhards** [comment]
> On Thu, Jun 14, 2018 at 04:14:30AM -0700, Rainer Gerhards wrote:
>  I understand that you just want to start a program when rsyslog starts up. It's not really meant to receive messages from rsyslog? Is that right?
> In my case yes.

Well... omprog is not meant for such a thing. I am not sure if you run into troubles. I guess when the program does not read from the pipe, rsyslog will at least emit error messages once it gets full. But I would say it's really the wrong tool for the job. 

> I'm not use systemd :-). But yes, i can make an init script with dependency
> of rsyslog and iptables. And in this script i would place a "kill -USR1 `cat /var/run/my.pid`"
> which switch the program already run by rsyslog to working mode.

well, then rsyslog will re-spawn the program, as it detects it died.

I think I will close the issue tracker now as this is not something within the desired rsyslog feature set. However, I am still open to good arguments why we need that functionality in here, happy to re-open then.

**Comment by lego12239** [comment]
On Thu, Jun 14, 2018 at 05:23:01AM -0700, Rainer Gerhards wrote:
> > On Thu, Jun 14, 2018 at 04:14:30AM -0700, Rainer Gerhards wrote:
> >  I understand that you just want to start a program when rsyslog starts up. It's not really meant to receive messages from rsyslog? Is that right?
> > In my case yes.
> 
> Well... omprog is not meant for such a thing. I am not sure if you run into troubles. I guess when the program does not read from the pipe, rsyslog will at least emit error messages once it gets full. But I would say it's really the wrong tool for the job.

  I incorrectly expressed. Sorry. I mean, when the program start it discard
all input from rsyslog. It _read_ it, but not processed. It wait when a
special event is occured(iptables is initialized in my case) and after this
it not only read input, but do input processing.

> > I'm not use systemd :-). But yes, i can make an init script with dependency
> > of rsyslog and iptables. And in this script i would place a "kill -USR1 `cat /var/run/my.pid`"
> > which switch the program already run by rsyslog to working mode.
> 
> well, then rsyslog will re-spawn the program, as it detects it died.

  Why do it must die?

-- 
Олег Неманов (Oleg Nemanov)

**Comment by rgerhards** [comment]
> I incorrectly expressed. Sorry. I mean, when the program start it discard
> all input from rsyslog. It _read_ it, but not processed. It wait when a
> special event is occured(iptables is initialized in my case) and after this
> it not only read input, but do input processing.

OK

do you pass it all messages or just a select subset? If so, why does it need to start as early as possible and not when the first interesting event comes in?

>  > I'm not use systemd :-). But yes, i can make an init script with dependency
>  > of rsyslog and iptables. And in this script i would place a "kill -USR1 `cat /var/run/my.pid`"
>  > which switch the program already run by rsyslog to working mode.
> 
>  well, then rsyslog will re-spawn the program, as it detects it died.
> Why do it must die?

my fault, I think I misunderstood

**Comment by lego12239** [comment]
On Thu, Jun 14, 2018 at 06:05:49AM -0700, Rainer Gerhards wrote:
> do you pass it all messages or just a select subset?

On different servers i use different settings.

> If so, why does it need to start as early as possible and not when the first interesting event comes in?

Because my program do some initialization on startup and i cann't
guarantee the order of init script on all distro/servers. That is,
i need the order - (rsyslog,iptables,ANY_OTHER_DEPENDANCY,my_prog_init).
Where my_prog_init is just an execution of init() function of my program.
Actually an order can be (rsyslog,my_prog_init,iptables,ANY_OTHER_DEPENDANCY)
or (rsyslog,iptables,my_prog_init,ANY_OTHER_DEPENDANCY); and this cause
a malfunction.

Since my_prog is run by rsyslog in any time later after rsyslog startup,
i can't guarantee my_prog_init execution in the right order.

So, i can toggle my_prog_init execution on some external event(i.e. on
a signal receiving), instead of execution of my_prog_init on my_prog startup.
And this signal can be sent from some init script with dependancy order
specified like this (rsyslog,iptables,ANY_OTHER_DEPENDANCY,my_prog_init_script).
But for this to work i need already running my_prog. Or i send signal to nowhere.

While i wrote this i got an idea how to workaround this without a signal :-).
I can simply do the next from my_prog_init_script:

touch /var/run/my_prog.run
pkill my_prog

And on startup check /var/run/my_prog.run existance and if it exist execute
my_prog_init().

But in any case, some option for omprog like is discussed here would be convenient.


Thanks!

-- 
Олег Неманов (Oleg Nemanov)

**Comment by davidelang** [comment]
On Thu, 14 Jun 2018, Oleg wrote:

> On Thu, Jun 14, 2018 at 06:05:49AM -0700, Rainer Gerhards wrote:
>> do you pass it all messages or just a select subset?
>
> On different servers i use different settings.
>
>> If so, why does it need to start as early as possible and not when the first interesting event comes in?
>
> Because my program do some initialization on startup and i cann't
> guarantee the order of init script on all distro/servers. That is,
> i need the order - (rsyslog,iptables,ANY_OTHER_DEPENDANCY,my_prog_init).
> Where my_prog_init is just an execution of init() function of my program.
> Actually an order can be (rsyslog,my_prog_init,iptables,ANY_OTHER_DEPENDANCY)

Well, if ryslog were to start it, the order could be

rsyslog, my_prog_init, iptables, anything else

or if your init system just starts rsyslog then continues, you still have race 
conditions because there is always going to be some variable amount of time 
after rsyslog starts before it can start anything else.

rsyslog has _no_ way of detecting that other programs have started, so it has no 
way of detecting your other dependencies.

So it seems to me that rsyslog starting your progrm sooner is not what you want. 
Instead you need to have a dependency on your other dependencies that rsyslog 
knows nothing about.

that's the job of your init system, not rsyslog.

David Lang

**Comment by lego12239** [comment]
David, you didn't understand me :-).

> Well, if ryslog were to start it, the order could be
>
> rsyslog, my_prog_init, iptables, anything else

No. The order will:

rsyslog, my_prog, iptables, OTHER_DEPENDANCY, my_prog_init_script

where my_prog_init_script can be(very simplified):

```
while ! pgrep my_prog; do sleep 1s; done   # wait my_prog if not started yet
pkill -USR1 `cat /var/run/my.pid`
```

**Comment by lego12239** [comment]
Forgot to say, when my_prog receive SIGUSR1, it executes my_prog_init().

**Comment by davidelang** [comment]
On Fri, 15 Jun 2018, Oleg wrote:

> David, you didn't understand me :-).
>
>> Well, if ryslog were to start it, the order could be
>>
>> rsyslog, my_prog_init, iptables, anything else
>
> No. The order will:
>
> rsyslog, my_prog, iptables, OTHER_DEPENDANCY, my_prog_init_script

how would rsyslog know to wait for iptables, etc before running your stuff? I 
thought you needed iptables started before my_prog?

> where my_prog_init_script can be(very simplified):
>
> ```
> while ! pgrep my_prog; do sleep 1s; done   # wait my_prog if not started yet
> pkill -USR1 `cat /var/run/my.pid`
> ```
>
>
>
>

**Comment by lego12239** [comment]
> how would rsyslog know to wait for iptables, etc before running your stuff?

rsyslog needn't know about it. Look, the order of init scripts:

1. rsyslogd
2. iptables
3. OTHER_DEPENDANCY
4. my_prog_init_script (wait my_prog to start and send SIGUSR1 to it)

Corresponding actions:

1. run my_prog with help of omprog
2. -
3. -
4. my_prog_init() is executed (my_prog receives SIGUSR1 from my_prog_init_script)

rsyslogd needn't know about anything related to my_prog. All actions are done by my_prog itself and my_prog_init_script. All that rsyslog need to do to this scheme to work, just start my_prog at startup.

**Comment by lego12239** [comment]
> thanks to parallel startup, rsyslog may still be processing it's config file at the time that the other startup  scripts run. Even with the older init systems, rsyslog is just started in the background, nothing watches it to figure out when it's ready to start processing log messages.

> So you still can have your other things startup before my_prog is runable and able to handle signals.

David, you forgot the example of my_prog_init_script i wrote above. All needed synchronization with my_prog startup i do myself. This is not the task or problem of rsyslog.

> I just don't think that having such a feature will give you the guarantee that you are looking for.

This is a good feature that will make rsyslog an even more better. If it will be implemented, i will make needed guarantee myself.

**Comment by jsiwrk** [comment]
I've not fully read the conversation, but just wanted to note that this feature (start external program on startup) could be easily supported when omprog is refactored to correct issue #2813. When the `forceSingleInstance` flag is enabled, the natural place to keep the data of the child process (running state, pid, FDs...) is the action instance. The child can then be started in [createInstance](https://github.com/rsyslog/rsyslog/blob/v8.36.0/plugins/omprog/omprog.c#L660) instead of the first received message. Actually, this will simplify the implementation I think. It would be symmetric with the case when `forceSingleInstance` is disabled: the child is started in [createWrkrInstance](https://github.com/rsyslog/rsyslog/blob/v8.36.0/plugins/omprog/omprog.c#L674) in this case.

**Comment by jsiwrk** [comment]
Closing as this feature is available since v8.38 by activating the forceSingleInstance flag (see my previous comment). This flag now not only guarantees that a single instance of the external program is started, but also that the instance is started at rsyslog startup time (and closed at shutdown).
